src/app/routes/api/v3/admin/forms/__tests__/admin-forms.twilio.routes.spec.ts

```diff
@@ -0,0 +1,341 @@
+import { PutSecretValueResponse } from 'aws-sdk/clients/secretsmanager'
+import { ObjectId } from 'bson-ext'
+import mongoose from 'mongoose'
+import { errAsync, okAsync } from 'neverthrow'
+import supertest, { Session } from 'supertest-session'
+
+import getFormModel from 'src/app/models/form.server.model'
+import getUserModel from 'src/app/models/user.server.model'
+import {
+  DatabaseError,
+  MalformedParametersError,
+} from 'src/app/modules/core/core.errors'
+
+import {
+  createAuthedSession,
+  logoutSession,
+} from 'tests/integration/helpers/express-auth'
+import { setupApp } from 'tests/integration/helpers/express-setup'
+import dbHandler from 'tests/unit/backend/helpers/jest-db'
+
+import * as AdminFormService from '../../../../../../../app/modules/form/admin-form/admin-form.service'
+import { AdminFormsRouter } from '../admin-forms.routes'
+
+import { TwilioCredentials } from './../../../../../../services/sms/sms.types'
+
+// Prevent rate limiting.
+jest.mock('src/app/utils/limit-rate')
+
+const app = setupApp('/admin/forms', AdminFormsRouter, {
+  setupWithAuth: true,
+})
+
+const UserModel = getUserModel(mongoose)
+const FormModel = getFormModel(mongoose)
+
+describe('admin-form.twilio.routes', () => {
+  let request: Session
+
+  beforeAll(async () => await dbHandler.connect())
+  beforeEach(async () => {
+    request = supertest(app)
+  })
+  afterEach(async () => {
+    await dbHandler.clearDatabase()
+    jest.restoreAllMocks()
+  })
+  afterAll(async () => await dbHandler.closeDatabase())
+
+  describe('PUT /admin/forms/:formId/twilio', () => {
+    const MOCK_FORM_ID = new ObjectId().toHexString()
+
+    const MOCK_ACCOUNT_SID = 'AC12345678'
+    const MOCK_API_KEY_SID = 'SK12345678'
+    const MOCK_API_KEY_SECRET = 'AZ12345678'
+    const MOCK_MESSAGING_SERVICE_SID = 'MG12345678'
+
+    const TWILIO_CREDENTIALS: TwilioCredentials = {
+      accountSid: MOCK_ACCOUNT_SID,
+      apiKey: MOCK_API_KEY_SID,
+      apiSecret: MOCK_API_KEY_SECRET,
+      messagingServiceSid: MOCK_MESSAGING_SERVICE_SID,
+    }
+
+    const MOCK_INVALID_ACCOUNT_SID = 'ZZ12345678'
+
+    const INVALID_TWILIO_CREDENTIALS: TwilioCredentials = {
+      accountSid: MOCK_INVALID_ACCOUNT_SID,
+      apiKey: MOCK_API_KEY_SID,
+      apiSecret: MOCK_API_KEY_SECRET,
+      messagingServiceSid: MOCK_MESSAGING_SERVICE_SID,
+    }
+
+    const MOCK_SUCCESSFUL_UPDATE = {
+      message: 'Successfully updated Twilio credentials',
+    }
+
+    it('should return 200 on successful twilio credentials addition', async () => {
+      const { form: formToUpdate, user } = await dbHandler.insertEmailForm()
+      const session = await createAuthedSession(user.email, request)
+
+      const twilioCredentialsSpy = jest
+        .spyOn(AdminFormService, 'createTwilioCredentials')
```

Comment:
we might want to mock deeper (and only external calls) instead of the service, and only mock AWS secrets manager so we can get assurance that the service works as intended

Reply:
I took a look at that the other controller unit tests, it seems that the other unit tests mocks the service level functions (I guess its to prevent errors from originating from the service functions). 

Just wondering if it will be better to keep it that way? As the admin-service unit tests is also there to ensure the admin-form service performs properly

Reply:
this is a route integration test and not a controller unit test, so we want to mock as little as possible actually. mocking secret manager only because it's an external API call